File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.py
from collections import deque
class Solution:
    def longestValidParentheses(self, s: str) -> int:
        d = deque([-1])
        answer = 0

        for i in range(len(s)):
            if s[i] == '(':
                d.append(i)
            else:
                d.pop()
                if len(d) == 0:
                    d.append(i)
                else:
                    answer = max(answer, i-d[-1])
        return answer    

File: 0032-longest-valid-parentheses/0032-longest-valid-parentheses.java
class Solution {
    public int longestValidParentheses(String s) {
        int answer = 0;
        int len = 0;
        Stack<Integer> stack = new Stack<>();
        stack.push(-1);
        
        for (int i = 0; i < s.length(); i++){
            if (s.charAt(i) == '('){
                stack.push(i);
            } else{
                if(stack.empty()){
                    continue;
                }
                stack.pop();
                if (stack.empty()){
                    stack.push(i);
                }
                else{
                    answer = (answer > i - stack.peek()) ? answer :  i - stack.peek();
                }
            }
        }
        return answer;
    }
}